tools/tool-testing/selftest.js

```diff
@@ -1856,8 +1868,108 @@ class TestList {
   // Mark a test's file as having failures.  This prevents
   // saveTestState from saving its hash as a potentially
   // "unchanged" file to be skipped in a future run.
-  notifyFailed(test) {
+  notifyFailed(test, failure) {
     this.fileInfo[test.file].hasFailures = true;
+    test.failure = failure || true;
+  }
+
+  saveJUnitOutput(path) {
+    const grouped = groupTestsByFile(this.filteredTests);
+
+    // We'll form an collection of "testsuites"
+    const testSuites = [];
+
+    const attrSafe = attr => (attr || "").replace('"', "&quot;");
+    const durationForOutput = durationMs => durationMs / 1000;
+
+    // Each file is a testsuite.
+    Object.keys(grouped).forEach((file) => {
+      const testCases = [];
+
+      let countError = 0;
+      let countFailure = 0;
+
+      // Each test is a "testcase".
+      grouped[file].forEach((test) => {
+        const testCaseAttrs = [
+          `name="${attrSafe(test.name)}"`,
+        ];
+
+        if (test.durationMs) {
+          testCaseAttrs.push(`time="${durationForOutput(test.durationMs)}"`);
+        }
+
+        const testCaseAttrsString = testCaseAttrs.join(' ');
+
+        if (test.failure) {
+          let failureElement = "";
+
+          if (test.failure instanceof TestFailure) {
+            countFailure++;
+
+            failureElement = [
+              `<error type="${test.failure.reason}">`,
+              '<![CDATA[',
+              inspect(test.failure.details, { depth: 4 }),
+              ']]>',
+              '</error>',
+            ].join('\n');
+          } else {
+            countError++;
+
+            failureElement = [
+              '<failure>',
+              '<![CDATA[',
+              test.failure.stack,
+              ']]>',
+              '</failure>',
+            ].join('\n');
+          }
+
+          testCases.push(
+            [
+              `<testcase ${testCaseAttrsString}>`,
+              failureElement,
+              '</testcase>',
+            ].join('\n'),
+          );
+        } else {
+          testCases.push(`<testcase ${testCaseAttrsString}/>`);
+        }
+      });
+
+      const testSuiteAttrs = [
+        `name="${file}"`,
+        `tests="${testCases.length}"`,
+        `failures="${countFailure}"`,
+        `errors="${countError}"`,
+        `time="${durationForOutput(this.durationMs)}"`,
+      ];
+
+      const testSuiteAttrsString = testSuiteAttrs.join(' ');
+
+      testSuites.push(
+        [
+          `<testsuite ${testSuiteAttrsString}>`,
+          testCases.join('\n'),
+          '</testsuite>',
+        ].join('\n'),
+      );
+    });
+
+    const xmlHeader = '<?xml version="1.0" encoding="UTF-8"?>';
+
+    const testSuitesString = testSuites.join('\n');
+
+    files.writeFile(path,
+      [
+        xmlHeader,
+        `<testsuites>`,
+        testSuitesString,
+        `</testsuites>`,
+      ].join('\n'),
+      'utf8',
+    );
```

Comment:
I know.  I almost did the isopacket approach (and introduced a new one, since `xmlbuilder` is only otherwise built into `cordova-support` and I secretly envisioned that going away and didn't like the idea of adding another isopacket.  Additionally, the `xmlbuilder` package used the `xmlbuilder` which was 7 major versions behind.